### PR TITLE
Fix Add and Subtract test failures in 330db and 340 shim [databricks]

### PR DESCRIPTION
**Pull request by razajafri**
This PR addresses the overflow that can happen when adding subtracting large numbers before rescaling the number. 

GpuRemainder and GpuPmod extend respective base classes and also override the `resultDecimalType` method which is needed for Spark 3.4 and DB11.3

Before 
`338 failed, 464 passed, 14 skipped, 2 xfailed in 448.22s (0:07:28)` 

After
`316 failed, 486 passed, 14 skipped, 2 xfailed in 444.10s (0:07:24)`

Contributes towards #7348  

Signed-off-by: Raza Jafri <rjafri@nvidia.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala [hunk 1]
```diff
@@ -175,92 +251,10 @@ case class GpuAbs(child: Expression, failOnError: Boolean) extends CudfUnaryExpr
...
-        withResource(Scalar.fromNull(ret.getType)) { nullVal =>
-          overflow.ifElse(nullVal, ret)
-        }
-      }
-    }
-  }
-}
-
-case class GpuAdd(
+abstract class GpuAddParent(
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala:254]: nit: suffix Base is used much more frequently in our code, also helps avoid confusion with plan node parent.
```suggestion
abstract class GpuAddBase(
```

Comment (razajafri) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/arithmetic.scala]: Done

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
@gerashegalov @revans2 Can you please take a look I think I have addressed your concerns

**Review by gerashegalov — APPROVED** [review]
LGTM

**Pull request merged**